### ISSUE TITLE
Full Co-add small area validation test idea list

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/DC2-production#320]
From slack:

We are currently considering a proposal where we would simulate a very small piece of the sky (around the size of a tract or a raft) to full 10 year depth for a validation test very soon.  We could do this simulation quite quickly as it does not use many sensors.  Also these simulations will not be “wasted”.  We have to eventually simulate these sensors in these visits anyway.

(see https://lsstc.slack.com/archives/C77DDKZHR/p1548777428203700 for more)

Details of test can be defined here:

Ideas for tests:

1. Check color of co-added galaxies (need at least two bands)
2. Check completeness of sample for faint objects. (number counts etc)
3. Ellipticity distributions
4. Compute total 5-sigma depth (DC1 made it to the minimum set by the LSST-SRD requirements but just barely although individual visits seemed to be within the margins, this test is also related with 2 above).

**Comment by fjaviersanchez** [comment]
4. Compute total 5-sigma depth (DC1 made it to the minimum set by the LSST-SRD requirements but just barely although individual visits seemed to be within the margins, this test is also related with 2 above).

**Comment by rmandelb** [comment]
I have some questions about the above tests; I guess the high-level concern I have is that I'm not sure these would be tests of the sims vs. of the DM stack and/or that they might not be covered by other tests.  Some specifics:

> Check color of co-added galaxies (need at least two bands)

Let's say that we do this and  make the coadds and find wonky colors, is that really a validation test of the sims?  What if it's because some PSF models or something in the coadd process or deblending/cmodel flux measurement  is going wrong?   I am concerned this could become a very lengthy process that ends up with multiple potential contributors that would take a while to disambiguate, and the majority of which are not simulation validation issues.

Right now I presumed our color validation test would be individual exposure photometry tests in >1 band, comparing truth catalogs vs. the DM outputs  (which we're already doing).  So I gather your concern is the faint galaxies with approximations applied to them, which aren't captured by such a test.  Is that correct? Could we consider doing something like generate an individual visit or two where we tell ImSim to have an extremely low sky background, so we can get into the regime where we are making the faint galaxy approximations that we presumably want to test?  Could we test their colors in some other way?

> Check completeness of sample for faint objects. (number counts etc)

**Comment by cwwalter** [comment]
> So I gather your concern is the faint galaxies with approximations applied to them, which aren't captured by such a test.

Right.. exactly.  But, good point on whether we will be able to know what the issue is given our lack of experience using the pipeline in this case.  

For this one the specific concern is that for very dim objects we are applying (for example) a simple SED instead of the full SED.  But, we aren't really even doing that.  Each galaxy component is handled differently. So for cases with (e.g) very dim disks those are being entered with a simple SED even if the total galaxy isn't dim.  So, it's pretty hard to see this effect for a single object/exposure as the flux is low on that component by definition.  The thing we are interested in checking is when you actually add them them all together and the S/N is good enough you can see those components if this effect (and the other similar speed up options) cause you not to represent things as you might have expected from the input.

**Comment by salmanhabib** [comment]
How about first simulating a handful of test objects instead where you can do detailed (but perhaps idealized) tests on a workstation? That is, if someone interested in doing this can be found --

**Comment by rmandelb** [comment]
I think I agree with @salmanhabib - but let me outline my reasoning in case there is a gap in the logic.  As I understand it, the SED approximations you mentioned are *preserving the total per-band flux* (and hence the u-g, g-r, etc. broad-band colors and the per-band number counts) and simply *approximating the within-band* SED shape.  Is that correct?  If it's not, then the rest of this comment is nonsense.  If it's correct, then please read on.

The approximation of the within-band SED basically means that certain color-dependent higher-order effects (such as DCR, chromatic PSFs, ...) will be *slightly* wrong for these very faint galaxies or for individual galaxy components that are faint enough to be subject to this approximation.  Right?

If so, then there are a few potential things that I think one could worry about:

1. _Actual implementation bugs_: i.e., due to a bug, something goes wrong and the broad-band colors aren't preserved, screwing up the number counts per band and so on.  Clearly, we should be doing unit tests or simple controlled tests with individual objects to ensure there is no such bug.   Were the tests that were already done adequate in this regard?  If not, might this be something we can check without simulating 10 years of data, but rather by inserting diagnostic statements (e.g., force it to tell us what flux is actually supposed to go into the image after simplifying the SED, and make sure it agrees with the case where we don't enable the SED approximations) or using simple test cases without any sky noise etc.?

2. _Design flaws in how this was implemented_: i.e., it is doing what we tell it to, but we're setting the flux thresholds wrong and accidentally approximating components of galaxies that are well-detected.  I presume this was already tested?  If more checks are needed, then this again seems like something that doesn't require simulation of 10 years of images.

3. _Low-level issues with the faint galaxy population due to the approximation_: i.e., that the impact of these higher-order chromatic effects due to within-band SED approximations does something to the measured sizes/shapes/colors on the final coadd.  

I am disinclined to worry about this, for the following reason: even the low-order aspects of the faint galaxy component light profiles, such as the broad-band colors, number counts, sizes, shapes, are based on sensible extrapolations of what we know from brighter populations.  The faint galaxy number counts can only be validated with pencil beam surveys to the 10-30% level to ~28th magnitude, and the sizes/shapes aren't validated at all. These seem like they would be actually *more* important effects than slight issues with the chromatic PSF or DCR etc., and we've already accepted them as approximations because there's no way to do better and because we think it's still better to have those faint galaxies with the associated question marks than to not have them.  Once that's the case, it seems inconsistent to me to worry about (3).

Thoughts on the above are welcome - I realize I haven't spelled this out before and could be missing something - but the bottom line is that none of 1-3 seem to me to call for simulating to 10 year depth.

To be clear, I am actually a fan of end-to-end integration tests and upping our game in terms of validation frameworks for the future, but I don't think this particular question is a good justification for an end-to-end integration test.

**Comment by cwwalter** [comment]
> The SED approximations you mentioned are preserving the total per-band flux (and hence the u-g, g-r, etc. broad-band colors and the per-band number counts) and simply approximating the within-band SED shape. Is that correct?

Yes, this is right.  The particular color suggestion was from Mike in our in person meeting, so I will let him say more.  I personally was thinking more about the morphology and components etc.

But, following up on the same of the comments above, I think the reason this (and the other general tests) are hard to do with say single objects on no background is because exactly what we decide to do in the algorithm (use a simple SED, drop the component etc etc) actually depends on the background level you are sitting on and what is around.  To be clear I don't think you need millions of objects. I think a single sensor or two would  be fine so the entire simulation could be done in less than a day, but I suspect we won't really know if this all works correctly unless we actually do it.  Of course, every time we do a careful check we find issues and I think it is probably easier to test the real system then to try to mock up things that approximate it.  OTOH, I think if doing the processing is really difficult or we don't know how to do it yet then this doesn't make sense at this point.

**Comment by cwwalter** [comment]
Oops forgot to add above: from the simulation perspective the cases where we could be doing things in a way that cause problems are all by definition the ones we can't really see in the single exposures since those are the galaxy components that are to dim to see in that case.  Only when we add them all together will we actually be able to see what the results are.  The components may not be handled consistently among exposures because of differing background levels etc.  So, this is what is driving my desire to see it all works.

**Comment by jchiang87** [comment]
With regard to checking the galaxy colors, I think there may already be sufficient information in the centroid files for the existing Run2.1i data.  We just need to pick a part of the sky that is sufficiently deep in a couple of band, and look at the realized photon counts for each type of rendered object, summed over visits and perhaps selecting or disaggregating further based on SED properties, etc..  We have a lot of faint galaxies to do this with, so ~30 visits deep may be enough.

**Comment by rmandelb** [comment]
> Oops forgot to add above: from the simulation perspective the cases where we could be doing things in a way that cause problems are all by definition the ones we can't really see in the single exposures since those are the galaxy components that are too dim to see in that case.

Right, I understand this.  My point was that if this approximation modifies the very faint galaxy number counts slightly, or the galaxy sizes/shapes/broad-band colors/morphology, I'm not worried about it because I don't trust the inputs for those quantities to better than 30% (number counts/SED details) or a factor of 2 (everything else I listed) anyway.  I mean, I trust that if this scheme is correctly implemented then the within-band SED approximations should affect those quantities at a far lower level than that, and hence are subdominant to the far larger and currently unreducable uncertainties due to the faint population inputs. So that's why I'm saying I am only concerned about making sure it's a bug-free implementation.  Obviously if there is a bug then potentially all bets are off and I'd easily believe that a bug in this approximation could lead to oddities that meet/exceed other uncertainties in the input population.  And then I suspect there is a way to find such bugs using some trick like what Jim suggested, or a special-order simulation that doesn't require the simulation of 10 years of data (for which it would be hard anyway to disambiguate DM vs simulation issues in the analysis of the final coadd).

**Comment by cwwalter** [comment]
OK, if people don't think the tests will be useful, so be it; we shouldn't waste resources on them then.  

FWIW, I am probably less worried about an outright bug, which I think we have probably done a pretty good job of checking for at this point (although you can never be sure) and more of what you might think of as unanticipated consequences and oversights coming from the speed up short cuts we took that we won't be able to see until the really dim individual galaxy components becomes visible.
 
I've had so many bad experiences after we didn't check everything first before months longs runs in the past that I will probably continue to work on this in parallel myself.  I can see this small data set would probably be useful for lots of things in any case (like we saw in the discussion in slack yesterday) so no matter what we can probably learn things from it.

**Comment by katrinheitmann** [comment]
@cwwalter Can we close this? Thanks!

**Comment by cwwalter** [comment]
@katrinheitmann I'm still working on this but, since we decided not to do these checks officially or before starting, I can track the work somewhere else.  I'll go ahead and close it.
